idl-parser/src/ffi/ast/mod.rs

```diff
@@ -1,18 +1,43 @@
 use crate::ast;
-use std::{slice, str};
+use std::{
+    error::Error,
+    ffi::{c_char, CString},
+    slice, str,
+};
 
 pub mod visitor;
 
+#[repr(C)]
+pub enum ParseResult {
+    Success(*mut Program),
+    Error(*const c_char),
+}
+
 /// # Safety
 ///
 /// See the safety documentation of [`slice::from_raw_parts`].
 #[no_mangle]
-pub unsafe extern "C" fn parse_idl(idl_ptr: *const u8, idl_len: u32) -> *mut Program {
-    let idl = unsafe { slice::from_raw_parts(idl_ptr, idl_len.try_into().unwrap()) };
-    let idl = str::from_utf8(idl).unwrap();
-    let program = ast::parse_idl(idl).unwrap();
-    let program = Box::new(program);
-    Box::into_raw(program)
+pub unsafe extern "C" fn parse_idl(idl_ptr: *const u8, idl_len: u32) -> *mut ParseResult {
+    let idl_slice = unsafe { slice::from_raw_parts(idl_ptr, idl_len as usize) };
+    let idl_str = match str::from_utf8(idl_slice) {
+        Ok(s) => s,
+        Err(e) => return create_error(e, "validate IDL string"),
+    };
+
+    let program = match ast::parse_idl(idl_str) {
+        Ok(p) => p,
+        Err(e) => return create_error(e, "parse IDL"),
```

Comment:
Apparently, memory allocated for the error message will leak. Should we just return some static text?

Reply:
Dynamic string will provide more information about error, including line:col of where parsing error happened. I've changed free logic to deallocate result instead of just program

Reply:
Hmm... But rust doesn't provide safe ffi for enums with fields. You can achieve it via using an intermediate enum for tagging, but it doesn't seem nice. Would it be more conventional to introduce yet another method for getting last error which takes a user passed buffer for error details as freeing error message via a separate call doesn't look common practice either